src/CodeGen_D3D12Compute_Dev.cpp

```diff
@@ -855,101 +922,115 @@ void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::add_kernel(Stmt s,
+        // In D3D12/HLSL, only 32bit types (int/uint/float) are suppoerted (even
+        // though things are changing with newer shader models). Since there is
+        // no uint8 type, we'll have to emulate it with 32bit types...
         // This will also require pack/unpack logic with bit-masking and aliased
         // type reinterpretation via asfloat()/asuint() in the shader code... :(
-        internal_assert(op->type == UInt(8));
+        Type smem_type = op->type;
+        smem_type.with_bits(32);
         stream << "groupshared"
-               << " " << print_type(op->type)  // print_type(uint8) -> uint32
+               << " " << print_type(smem_type)
                << " " << print_name(op->name);
         if (is_const(op->extents[0])) {
             std::stringstream ss;
             ss << op->extents[0];
             size_t elements = 0;
             ss >> elements;
             size_t bytesize = elements * sizeof(uint32_t);
-            // SM 5.1: 32KB limit for shared memory...
-            size_t packing_factor = 1;
-            while (bytesize > 32 * 1024) {
-                // must pack/unpack elements to/from shared memory...
-                elements /= 2;
-                bytesize /= 2;
-                packing_factor *= 2;
-            }
+            // NOTE(marcos): might need to resort to StoragePackUnpack::pack_storage() here...
+            internal_assert(bytesize <= StoragePackUnpack::ThreadGroupSharedStorageLimit);
+            total_shared_bytes += bytesize;
             stream << " [" << elements << "];\n";
-            // smallest possible pack type is a byte (no nibbles)
-            internal_assert(packing_factor <= 4);
         } else {
             // fill-in __GROUPSHARED_SIZE_IN_BYTES later on when D3DCompile()
             // is invoked in halide_d3d12compute_run(); must get divided by 4
             // since groupshared memory elements have 32bit granularity:
             stream << " [ ( __GROUPSHARED_SIZE_IN_BYTES + 3 ) / 4 ];\n";
         }
+        if (total_shared_bytes > StoragePackUnpack::ThreadGroupSharedStorageLimit) {
+            debug(1) << "D3D12 CodeGen ERROR: Total thread group shared memory required for kernel '" << name
+                     << "' exceeds the SM 5.1 limit of 32KB: " << total_shared_bytes << " bytes required.\n";
+        }
         Allocation alloc;
         alloc.type = op->type;
         allocations.push(op->name, alloc);
     }
 
-    // Find and patch situations where the __shared buffer is read before having
-    // ever being initialized:
+    // Find and patch situations where threadgroup shared buffers are read before
+    // having ever being initialized:
 
     // NOTE(marcos): it would be cleaner if we could just use an IRVisitor here
     // but in order to find the enclosing Stmt of a Load expression we need to
     // to walk through base Stmt nodes, and only IRMutator has this overload:
     struct FindUninitializedSharedLoads : public IRMutator {
```

Comment:
What do you think of just replacing all loads to buffers where there hasn't been a prior store with zero of the appropriate type? It would be much simpler and could be done in a single mutator. When you encounter the load, if you haven't seen a store yet, just return zero instead.